### ISSUE TITLE
Introduce a priority system for auditing street edges

### ISSUE THREAD

**Issue by misaugstad** [ProjectSidewalk/SidewalkWebpage#1100]
Context: @manaswis can correct me if I'm wrong, but I believe that right now, when a user goes to the audit page, the front-end is sent all the street edges in that neighborhood, along with a count of how many times each has been audited. Then when the user reaches the end of a street, we see if it is connected to any least-audited street. If so, we route them down that street, otherwise we have them jump to a least-audited street.

By least-audited, I mean that if the neighborhood is not completely audited (across *all* users), then the user will only be sent down unaudited streets. If the neighborhood has been completed, the user will only go down streets that have been audited once. And so on.

I would argue that this is an implicit priority ranking for street edges, where auditing a street edge that has never been audited is the highest priority, followed by street edges that have only been audited once, and so on.

---

My proposal is that we extend this to be an explicit priority ranking. This would give us greater flexibility, by allowing us to define other sorts of priorities.

For example, the first thing that comes to mind is when we create some model of how reliable a label (or set of labels on a street edge) is, we may want to have another user (a "fresh pair of eyes", so to speak) audit the same street. Another example would be that we have detected that Google has uploaded more recent panoramas on a street, so it would be nice to have a new audit of that street.

So you could imagine a priority ordering such as...
1. Unaudited streets
2. Audited, but we suspect the audit may be of poor quality
3. Auidted, but panos have been updated since last audit
4. Been audited only once
5. Been audited only twice
...

This could also give us the flexibility on the front-end to give a better user experience with less jumping, by choosing to let the user go down any street that falls into one of the first 2 or 3 categories, instead of just the first category.

Any thoughts, or did I say anything confusing?

**Comment by jonfroehlich** [comment]
Thanks for starting this thread. I think there are some **very** interesting algorithmic research problems in here (that could possibly be a standalone paper as well). Soheil and I spoke with John Dickerson about it one time and I had some extensive conversations with Kotaro too. 

Two additional aspects that I think are worth mentioning:

1. One is some a priori prediction about the accessibility of a street based on a learned model from our prior work. The model would take, as input, things like real estate prices, street types, land use, building age, and possibly other open datasets (e.g., curb ramp data from open gov websites) and then spit out a prediction of accessibility with the idea that we would want to audit low accessibility places first.

2. Second, is an initial resource allocation optimization for getting first-pass coverage of a neighborhood: here, I was thinking something like to get some initial model of the accessibility of a neighborhood, it's perhaps not necessary to audit every single street but something like every other street (or whatever). Once we finish gathering data in DC, we can investigate this offline and see how well the accessibility of one street predicts another, etc. (but something probably more complex than this). Imagine being able to get a strong notion of neighborhood accessibility by auditing only 50% of the streets (thus, 2x savings!). Obviously, we would still want to get comprehensive data for some applications like route planning but I still think doing this is useful. This also brings to mind that everything we have in our system should be probabilistic: each label has a probability of being correct, each turker has a probability of work quality, etc.

The cool thing about this discussion is that I think there are broad implications for virtual auditing work in general, not just that applied to accessibility, but also public health, neighborhood safety & urban decay tracking,, tree coverage and assessment, etc. (and other uses of virtual auditing from literature).

**Comment by jonfroehlich** [comment]
We just briefly chatted about this during our weekly meeting, so I wanted to follow-up here about adding an additional column to the database for completed routes that is something like 'inferred_work_quality' that is some estimate of the quality of labels along that completed route for that user. The algorithm to make this determination is currently an open problem and was also discussed.

**Comment by misaugstad** [comment]
@maddalihanumateja and I had a nice discussion about this yesterday, to the point where we think we know what should be done next. I want to run the ideas by you @jonfroehlich and @manaswis 

Note that the ideas here about how how to provide a _priority system_ that helps to decide which streets should be audited next. Although this is partially about inferring work quality, we are not talking about the specifics of how to infer said work quality (that is, as you have said, a PhD-worthy subject). This is about how to design the _structure_ so that we can have a more sophisticated system that decides which streets need to be audited, such that we can start small, but the system is easily extendable.

For context, here is our current system for determining which streets a user should audit within a given neighborhood:
1. Count the number of completed audit tasks associated with each street segment in the neighborhood.
2. Find the minimum completion count, `minCount` among streets in the neighborhood (for a neighborhood that is not 100% audited, this would mean the min is 0).
3. Only route the user down streets that have a completion count of `minCount`.

This has been serving us pretty well so far; when the neighborhood is incomplete, users are only sent down unaudited streets, then once the neighborhood is fully audited, users are only sent down streets that have been audited once.

The problem with this approach: We want to take a lot more into account when choosing a street to route a user through. Some example we mentioned above: How reliable were the users that audited this street in the past? Have new panoramas been added to this street segment since it was last audited? How long ago was this street audited? There are many others that we might want to add.


### The new system we are proposing
Given the wide array of inputs we might want to have, our idea was to have some sort of priority "score", probably from 0-1, with 1 being highest priority. We would then calculate this priority score for a given street edge by combining any number input functions, with specified weights. This makes the system highly extendable, since we can always add in a new input function to make the system more precise.

#### Example
For example, consider our current situation, where our only input is the number of completed audits. We could define our input function for number of audits as `1 - x/(x+1)`. So streets with 0 completed audits would get a priority 1, streets with 1 audit a priority of 0.5, then 0.33, and so on. Since this is our only input right now, this particular function would be given full weight.

Now let's say that we want to add in a new input: the confidence we have in workers that have labeled. So let's say the @maddalihanumateja created a simple classifier that is based on the speed that a worker audits and the number of labels that they place, and that this classifier puts a worker into one of 3 categories: high confidence, medium confidence, low confidence in their work. So let's define a function that uses this classification of each user to get a priority score for each street edge:

For a given street edge e, it has been audited by n workers, w<sub>1</sub>, w<sub>2</sub>, ..., w<sub>n</sub>, each with an associated confidence level in their work c<sub>1</sub>, c<sub>2</sub>, ..., c<sub>n</sub> (let's say that high confidence has a score of 1, medium confidence has a score of 0.5, and low confidence has a score of 0). Since this is a simple example, let's just assign a priority score for the edge, e, as max(c<sub>i</sub>). Finally for the special case where the edge is unaudited, let's have this function give a priority score of 0.

So now we have two input functions, one for the number of completed audits (with score s<sub>1</sub>), and one for the confidence we have in the auditors for that street edge (with score s<sub>2</sub>. Now we can just assign weights to each input, let's say 50/50. So we can compute the final priority score for a street edge as 0.5*s<sub>1</sub> + 0.5*s<sub>1</sub>.


I realize that there is a lot of text here, but the concept really is pretty simple; let me know if I should clarify anything. Hopefully it is clear that this is a system that is easily extendable, in a way that I think will serve future people working on the project who want to implement more complicated quality control measures. This system also seems really simple to implement right now, even into our current system.

@jonfroehlich let me know what you think about this sketch, and @maddalihanumateja let me know if you think I misrepresented anything that we talked about. I am happy to talk about more implementation details, I have ideas of how to do most of it, but I wanted to keep this initial description high level :)

**Comment by jonfroehlich** [comment]
In terms of a basic approach for inferring labeler quality and mission/route quality, I propose something like this (I've proposed something similar for the last 12 months but can't seem to find any written documentation at the moment).

1. Brainstorm discrete features that can be fed into an ML algorithm (the “feature engineering” part of the problem). This would include both low-level interaction features (e.g., mouse speed, keypresses, etc.) as well as high-level, application-related features (e.g., labeling rates, panos visited, walking speed, mission features).
2. Use scikit-learn or some other off-the-shelf ML framework to train an algorithm using these features. I think we could use an SVM. Use 10-fold cross validation on the GT data to help validate/improve our approach.
3. Iterate on features and model that we use. Figure out which features provide the most information gain.
4. Apply this model to all data on production server. This ML model would run each night via a cron job (or something) and update new fields in the database like quality_inference in the worker table and quality_inference in the mission_complete (or street_segment) table (or whatever).

For potential features to use as input, see the documents in the Dropbox folder: Dropbox\ProjectSidewalk_CHI2018\BehavioralFeatures
